src/function/scalar/date/time_bucket.cpp

```diff
@@ -0,0 +1,348 @@
+			throw OutOfRangeException("Timestamp out of range");
+		}
+		ts_micros -= origin_micros;
+
+		int64_t result_micros = (ts_micros / bucket_width_micros) * bucket_width_micros;
+		ts_micros -= result_micros;
+		if (ts_micros < 0 && ts_micros % bucket_width_micros != 0) {
+			if (result_micros < NumericLimits<int64_t>::Minimum() + bucket_width_micros) {
+				throw OutOfRangeException("Timestamp out of range");
+			}
+			result_micros -= bucket_width_micros;
+		}
+		result_micros += origin_micros;
+
+		return Timestamp::FromEpochMicroSeconds(result_micros);
+	}
+
+	static date_t WidthMoreThanMonthsCommon(int32_t bucket_width_months, int32_t ts_months, int32_t origin_months) {
+		origin_months %= bucket_width_months;
+		if (origin_months > 0 && ts_months < NumericLimits<int32_t>::Minimum() + origin_months) {
+			throw NotImplementedException("Timestamp out of range");
+		}
+		if (origin_months < 0 && ts_months > NumericLimits<int32_t>::Maximum() + origin_months) {
+			throw NotImplementedException("Timestamp out of range");
+		}
+		ts_months -= origin_months;
+
+		int32_t result_months = (ts_months / bucket_width_months) * bucket_width_months;
+		if (ts_months < 0 && ts_months % bucket_width_months != 0) {
+			if (result_months < NumericLimits<int32_t>::Minimum() + bucket_width_months) {
+				throw OutOfRangeException("Timestamp out of range");
+			}
+			result_months = result_months - bucket_width_months;
+		}
+		result_months += origin_months;
+
+		int32_t year =
+		    (result_months < 0 && result_months % 12 != 0) ? 1970 + result_months / 12 - 1 : 1970 + result_months / 12;
+		int32_t month =
+		    (result_months < 0 && result_months % 12 != 0) ? result_months % 12 + 13 : result_months % 12 + 1;
+
+		return Date::FromDate(year, month, 1);
+	}
+
+	struct WidthLessThanDaysBinaryOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA bucket_width, TB ts) {
+			if (!Value::IsFinite(ts)) {
+				return Cast::template Operation<TB, TR>(ts);
+			}
+			int64_t bucket_width_micros = Interval::GetMicro(bucket_width);
+			int64_t ts_micros = Timestamp::GetEpochMicroSeconds(Cast::template Operation<TB, timestamp_t>(ts));
+			return Cast::template Operation<timestamp_t, TR>(
+			    WidthLessThanDaysCommon(bucket_width_micros, ts_micros, DEFAULT_ORIGIN_MICROS));
+		}
+	};
+
+	struct WidthMoreThanMonthsBinaryOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA bucket_width, TB ts) {
+			if (!Value::IsFinite(ts)) {
+				return Cast::template Operation<TB, TR>(ts);
+			}
+			int32_t ts_months = EpochMonths(ts);
+			return Cast::template Operation<date_t, TR>(
+			    WidthMoreThanMonthsCommon(bucket_width.months, ts_months, DEFAULT_ORIGIN_MONTHS));
+		}
+	};
+
+	struct BinaryOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA bucket_width, TB ts) {
+			if (bucket_width.months == 0) {
+				int64_t bucket_width_micros = Interval::GetMicro(bucket_width);
```

Comment:
I think the code would be cleaner if there was a method to validate bucket widths:

```cpp
static bool ValidateBucketWidth(const inteval_t &bucket_width);
```
It would return true for months, false for days and throw if it found a violation. Then the error messages would all be in once place along with the tests?

Reply:
I wrote `BucketWidthType ClassifyBucketWidth(const interval_t bucket_width)` and   `BucketWidthType ClassifyBucketWidthWithThrowingError(const interval_t bucket_width)` instead of `ValidateBucketWidth`.
The need to throw an error depends on whether it is before or after the null check, so use both.